frontend/src/hooks/useForm.ts

```diff
@@ -0,0 +1,55 @@
+import { useState, useCallback, useEffect } from "react";
+
+export interface FormErrors {
+  readonly [field: string]: string | undefined;
+}
+
+type Validator<V> = (values: V) => FormErrors;
+
+/**
+ * Custom hook for setting up a form with validation.
+ *
+ * @param initValues the initial values for each field
+ * @param validators the validator for the form
+ */
+export const useForm = <V>(initValues: V, validator: Validator<V>) => {
```

Comment:
Setting up a form requires specifying the initial values for each field in the form, along with a single validator function for creating an error object that contains the errors for each field.

The hook will give back several values which are used to then wire up the form, including the field values, a function for setting the values, the errors, a function for checking if there are any errors at all (used on submit), whether the errors should be visible, and a function for resetting the entire form.